pkg/controller/components/apiproxy.go

```diff
@@ -0,0 +1,166 @@
+	projectIDLabel             = "dominodatalab.com/project-id"
+	datasourceProxyClientLabel = "datasource-proxy-client"
+)
+
+func executionID(obj *client.Object) string {
+	if executionID, ok := (*obj).GetLabels()[executionIDLabel]; ok {
+		return executionID
+	}
+	return "unknown-execution-id" // primarily, for integration testing
+}
+
+func runPodSelector(obj *client.Object) map[string]string {
+	runPodName := fmt.Sprintf("run-%s", executionID(obj))
+	return map[string]string{
+		metadata.ApplicationInstanceLabelKey: runPodName,
+	}
+}
+
+func newResourceMeta(obj *client.Object, componentMeta *metadata.Provider) metav1.ObjectMeta {
+	instanceName := fmt.Sprintf("%s-%s", component, executionID(obj))
+	return metav1.ObjectMeta{
+		Name:      instanceName,
+		Namespace: (*obj).GetNamespace(),
+		Labels:    componentMeta.StandardLabelsWithComponent(*obj, component, nil),
+	}
+}
+
+func NewAPIProxyServiceComponent(obj *client.Object, port int32, meta *metadata.Provider) *corev1.Service {
+	var apiProxyPort int32
+	if port != 0 {
+		apiProxyPort = port
+	} else {
+		apiProxyPort = defaultAPIProxyPort
+	}
+
+	ports := []corev1.ServicePort{{
+		Name:       apiProxyPortName,
+		Port:       apiProxyPort,
+		TargetPort: intstr.FromInt(targetServicePort),
+		Protocol:   corev1.ProtocolTCP,
+	}}
+
+	internalTrafficPolicy := corev1.ServiceInternalTrafficPolicyCluster
+	ipFamilyPolicy := corev1.IPFamilyPolicySingleStack
+
+	return &corev1.Service{
+		ObjectMeta: newResourceMeta(obj, meta),
+		Spec: corev1.ServiceSpec{
+			Selector:              runPodSelector(obj),
+			Ports:                 ports,
+			InternalTrafficPolicy: &internalTrafficPolicy,
+			Type:                  corev1.ServiceTypeClusterIP,
+			IPFamilies:            []corev1.IPFamily{corev1.IPv4Protocol},
+			IPFamilyPolicy:        &ipFamilyPolicy,
+			SessionAffinity:       corev1.ServiceAffinityNone,
+		},
+	}
+}
+
+type APIProxyServiceComponent struct {
+	APIProxyPort func(obj *client.Object) int32
+	Meta         *metadata.Provider
+}
+
+func (c APIProxyServiceComponent) Reconcile(ctx *core.Context) (ctrl.Result, error) {
+	obj := ctx.Object
+	svc := NewAPIProxyServiceComponent(&obj, c.APIProxyPort(&obj), c.Meta)
+
+	err := actions.CreateOrUpdateOwnedResource(ctx, obj, svc)
+	if err != nil {
+		err = fmt.Errorf("cannot reconcile service: %w", err)
+	}
+
+	return ctrl.Result{}, err
+}
+
+func (c APIProxyServiceComponent) Kind() client.Object {
+	return &corev1.Service{}
+}
+
+type APIProxyNetworkPolicyComponent struct {
+	APIProxyPort func(obj *client.Object) int32
+	Meta         *metadata.Provider
+}
+
+func NewAPIProxyNetworkPolicyComponent(obj *client.Object, port int32, meta *metadata.Provider) *networkingv1.NetworkPolicy {
+	var apiProxyPort intstr.IntOrString
+	if port != 0 {
+		apiProxyPort = intstr.FromInt(int(port))
+	} else {
+		apiProxyPort = intstr.FromInt(defaultAPIProxyPort)
+	}
+
+	targetSelector := map[string]string{
```

Comment:
It's actually easier, in my view, to match by a combination `execution-id` and `datasource-proxy-client` labels, rather than via `name` and `instance`, because the latter pair always includes a cluster type name (`spark` etc). BTW, the `targetName` variable name is incorrect -- should be `sourceName`.

Reply:
But the DCO doesnt know about execution ids, or that the pods have data source proxy client labels. There are specific labels that it sets on worker pods so that it can create these selectors for its own use.

Reply:
OK, looks like my approach of parsing execution ID and using it to create labels was incorrect. (I keep forgetting that DCO is of a more general purpose than it seems, although at this point I am not completely sure this was a right idea).